### PR TITLE
Initial implementation of Plan 1 rechit combination

### PR DESCRIPTION
**Pull request by igv4321**
Initial implementation of Plan 1 rechit combination. Discussed at
https://indico.cern.ch/event/602307/contributions/2430302/attachments/1394590/2125473/notes.pdf

This PR is not supposed to modify any existing results. The testing
will be performed when the Plan 1 geometry becomes up to snuff.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by igv4321** [comment]
It looks like some input files are missing. The test failure is, of course, unrelated to this PR.

File: RecoLocalCalo/HcalRecProducers/python/HBHEPlan1Combiner_cfi.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import FWCore.ParameterSet.Config as cms
+
+hbheprereco = cms.EDProducer(
```
Comment (kpedro88) [inline RecoLocalCalo/HcalRecProducers/python/HBHEPlan1Combiner_cfi.py]: Can you define a unique name for this module?

**Comment by igv4321** [comment]
@kpedro88 
Well, I thought that you would do that, depending on what you want to rename. But OK, done.

**Comment by mdhildreth** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test
to capture all comparisons

**Comment by slava77** [comment]
@igv4321 @kpedro88 @bsunanda 
how far are we with the plan-1 geometry implementation so that this PR can be tested at run time?

File: RecoLocalCalo/HcalRecAlgos/interface/AbsPlan1RechitCombiner.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    virtual void combine(HBHERecHitCollection* collectionToFill) = 0;
+
+protected:
+    // The first element of the pair is the value to average
+    // and the second is the weight (energy). Non-positive weights
+    // will be ignored.
+    typedef std::pair<float, float> FPair;
+
+    static float energyWeightedAverage(const FPair* data, unsigned len,
+                                       float valueToReturnOnFailure);
```
Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/interface/AbsPlan1RechitCombiner.h]: the argument types better match the implementation

in .cc:
```c++
float AbsPlan1RechitCombiner::energyWeightedAverage(
    const FPair* data, const unsigned len,
    const float valueToReturnOnFailure)
```

File: RecoLocalCalo/HcalRecProducers/python/HBHEPlan1Combiner_cfi.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import FWCore.ParameterSet.Config as cms
+
+hbheplan1 = cms.EDProducer(
```
Comment (slava77) [inline RecoLocalCalo/HcalRecProducers/python/HBHEPlan1Combiner_cfi.py]: please keep the cfi file base and the instance defined in it the same

File: RecoLocalCalo/HcalRecProducers/src/HBHEPlan1Combiner.cc [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+private:
+    virtual void produce(edm::Event&, const edm::EventSetup&) override;
+
+    //virtual void beginStream(edm::StreamID) override;
+    //virtual void endStream() override;
+    //virtual void beginRun(edm::Run const&, edm::EventSetup const&) override;
+    //virtual void endRun(edm::Run const&, edm::EventSetup const&) override;
+    //virtual void beginLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override;
+    //virtual void endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override;
```
Comment (slava77) [inline RecoLocalCalo/HcalRecProducers/src/HBHEPlan1Combiner.cc]: is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

**Comment by slava77** [comment]
IIUC, this will need an era to be able to use the combined hits.
@kpedro88 do we have this era for plan-1?

**Comment by slava77** [comment]
@igv4321 
I'm done with the visual inspection.
Sorry, I didn't get around to this sooner.

**Comment by igv4321** [comment]
The argument types do match the implementation.
If you don't know why the const qualifiers are not used
in the method declaration, please read section 14.10,
"When should const not be used in declaring formal parameters?",
in the book "C++ FAQs, Second Edition" by Marshall Cline, Greg Lomow,
and Mike Girou.

**Comment by igv4321** [comment]
Do I really have to worry about this? Obviously,
the authors of "mkedprod" wanted to keep these
comments in. It makes sense to me -- in the future,
if I want to implement one of these functions,
I don't have to look up the interface definitions.

Now, imagine applying your comments to all
CMSSW module authors. Does it make sense?
Not really -- instead, one should modify "mkedprod".

**Comment by igv4321** [comment]
OK, renamed.

**Comment by igv4321** [comment]
I am done with replies and modifications. Please let me know if this works.

**Comment by slava77** [comment]
On 2/6/17 7:23 AM, Igor Volobouev wrote:
> The argument types do match the implementation.
> If you don't know why the const qualifiers are not used
> in the method declaration, please read section 14.10,
> "When should const not be used in declaring formal parameters?",
> in the book "C++ FAQs, Second Edition" by Marshall Cline, Greg Lomow,
> and Mike Girou.
> 

OK

**Comment by slava77** [comment]
On 2/6/17 7:41 AM, Igor Volobouev wrote:
> Do I really have to worry about this? Obviously,
> the authors of "mkedprod" wanted to keep these
> comments in. It makes sense to me -- in the future,
> if I want to implement one of these functions,
> I don't have to look up the interface definitions.
> 
> Now, imagine applying your comments to all
> CMSSW module authors. Does it make sense?
> Not really -- instead, one should modify "mkedprod".

The skeleton should contain all relevant methods.

The actual code doesn't really have to,
if it's not using them.

**Comment by kpedro88** [comment]
@igv4321 I have not started working on the era yet. I will try to get to it today. I would prefer to have this and #17330 merged, and then make the PR creating the Era/scenario. At that time, we can fix any bugs that show up.

File: RecoLocalCalo/HcalRecProducers/python/hbheplan1.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import FWCore.ParameterSet.Config as cms
+
+hbheplan1 = cms.EDProducer(
```
Comment (slava77) [inline RecoLocalCalo/HcalRecProducers/python/hbheplan1.py]: sorry for possibly unclear message earlier: _cfi should stay in the file name.
Thank you.

**Comment by igv4321** [comment]
OK, comments removed. Hopefully, this increases the overall
level of happiness.

**Comment by igv4321** [comment]
OK, done.

**Comment by igv4321** [comment]
I agree.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
for https://github.com/cms-sw/cmssw/pull/17313 	9d73ebf
- code changes are in line with the PR description and the follow up reivew
- jenkins tests pass and comparisons show no differences (the standard workflows are not modified by this PR)

- local test with ```hbheplan1.ignorePlan1Topology = True``` and ```hbheplan1.hbheInput = "hbhereco"``` on top of phase-2 wf 21234.0 shows that the module appears to work, but with some limitations. In this setup I would expect the hbheplan1 hits to be a copy of the hbhereco hits
    - the output hits have the same energy and chi2. Good.
    - timeFalling_ occasionally differs for energy=0 hits: hbhereco has 0 while hbheplan1 has -120 (the special value as encoded) (~OK?)
    - auxHBHE_ and aux_ are not set in hbheplan1 (bad?)
    - time() is generally different within floating point precision; cases with 0 energy have hbheplan1 time set to -999 (~ OK?)

@igv4321 please check and clarify if aux fields are to be abandoned/ignored or if there is a meaningful way to combine.
Are these used anywhere?

**Comment by igv4321** [comment]
Yes, all effects that you observe are intended.

hbhereco has 0 for timeFalling_ because timeFalling_ is used to store TDC time of QIE11 rechit. This time is not simulated in MC currently. The combined time will be driven to -120 in case either all QIE11 TDC times are "special" (underflow/overflow -- but this you can't observe with MC) or the rechit energy is not positive (no way to calculate energy-weighted TDC average).

auxHBHE_ and aux_ are not filled because, in the original rechits, they are used to store raw ADC counts. These counts are used for various diagnostic purposes. However, there appears to be no meaningful way to combine them.

hbheplan1 time is set to -999 if the energy is not positive. We probably don't care much about these rechits (and, again, there is no good way to calculate energy-weighted time average).

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17313 	9d73ebf
- see https://github.com/cms-sw/cmssw/pull/17313#issuecomment-277801206 and https://github.com/cms-sw/cmssw/pull/17313#issuecomment-277812302

**Comment by davidlange6** [comment]
+1

**Pull request merged**